### PR TITLE
GPS panel

### LINKED ISSUE #1679: GPS panel
A panel to show GPS status and accuracy. 

1- There could be two mode:
  - A summary mode to be displayed over the map 
image

  - A full mode to show more details
image

(screenshots are from GPS Status app: https://play.google.com/store/apps/details?id=com.eclipsim.gpsstatus2&hl=en)

2- The panel should support both internal GPS chip and external GPS

### LINKED ISSUE #1739: improve UX of the select active layer
during the training 3/14 students were unable to find in inputapp how to select the different layer for digitizing. It looks like the selector on the bottom of the screen with combination of small text is not visually too impressive so it is difficult to see

### PR DESCRIPTION
**Pull request by tomasMizera**
PR adds new GPS panel that is capable of showing current position, position in project's CRS, altitude, accuracy, number of used/viewed satellites, speed and timestamp of last fix (last time we received GPS position).

This panel can be opened via new map floating button that shows horizontal accuracy. 

**How it works:**
 - data are acquired from `positionkit.h`
 - `positionkit.h` has new properties to match needed GPS information
 - we now also use `QGeoSatelliteInfoSource` class to find information about satellites in use/satellites in view. This source is used only when real location is set, not simulated position.
 - we have new reusable component called `MapFloatButton` that can be reused in future for similar buttons
 - If GPS is not available, accuracy button is hidden, thus GPS page can not be opened!

**What has changed:**
 - GPS accuracy threshold setting has changed from numberspin (int values from 0 to 99) to custom field where user can type in any real number - in order to have cm accuracy threshold
 - Active layer panel was redesigned to match design of new floating button (resolves #1739)

Resolves #1679 

Some screenshots:
<table>
  <tr>
    <td></td>
    <td></td>
  </tr>
  <tr>
    <td></td>
    <td></td>
  </tr>
 </table>

Video to showcase the panel (I am changing my mocked location, accuracy and precision):

https://user-images.githubusercontent.com/22449698/141077802-98b9a4fc-2591-49b4-89de-c80619926bac.mp4

> Note: video contains old GPS data page layout, look at previous screenshots to get the most recent layout

### CODE REVIEW

**Comment by jancelin** [comment]
After a test with my GNSS RTK antenna, it works very well!
Thanks for the development!

photo_2021-11-10_11-22-23
photo_2021-11-10_11-22-09
photo_2021-11-10_11-22-05
photo_2021-11-10_11-22-04

**Comment by tomasMizera** [comment]
Glad to hear that @jancelin, thanks for testing it out with your devices! :)

**Comment by jancelin** [comment]
> Glad to hear that @jancelin, thanks for testing it out with your devices! :)

I think I see a problem on the vertical accuracy 
Screenshot_20211110-140005~2 Screenshot_20211110-135829~2

**Comment by tomasMizera** [comment]
Hi @jancelin, can you please test it when at least 4 satellites will be used? From your screenshot I see that there are no satellites (0/0).

**Comment by jancelin** [comment]
> Glad to hear that @jancelin, thanks for testing it out with your devices! :)



> Hi @jancelin, can you please test it when at least 4 satellites will be used? From your screenshot I see that there are no satellites (0/0).

I don't use the internal gps but an external one + a ntrip client for RTK which sends me back NMEA in bluetooth. I think it comes from there because I have at least 24 satellites used.

**Comment by jancelin** [comment]
internal smartphone GNSS  **VS** external GNSS RTK receiver

**Comment by tomasMizera** [comment]
Update: issues with vertical accuracy and satellites will be handled in separated ticket (https://github.com/lutraconsulting/input/issues/1753). 

GPS Page layout was updated based on @wonder-sk's requirements. @jozef-budac ready for testing

**Comment by jozef-budac** [comment]
Result of feature testing. There are some UI/UX problems:
1) in GPS panel, after pressing back button, application is closed. Map should be displayed
2) if layer name is too long, it renders outside of the floating button
3) when setting wrong number like "1.1.1" to the Accuracy threshold in settings, you can remove it just by selecting the whole text and remove it. It can't be deleted by remove button.

File: app/positionkit.h [hunk 1]
```diff
@@ -192,11 +230,25 @@ class PositionKit : public QObject
...
+    void stopUpdates();
+
     /**
      * Use real GPS source (not simulated)
      */
     Q_INVOKABLE void useGpsLocation();
 
+    const QDateTime &lastGPSRead() const;
+
+    QGeoSatelliteInfoSource *satelliteSource() const;
```
Comment (PeterPetrik) [inline app/positionkit.h:246]: what properties do you read from it? do we need to expose the source outside of the position kit? also , why the satellite source is separate from "source"... ?

File: app/inpututils.cpp [hunk 2]
```diff
@@ -236,6 +236,16 @@ double InputUtils::convertCoordinateString( const QString &rationalValue )
...
 }
 
+QString InputUtils::degreesString( const QgsPoint &point )
+{
+  if ( point.isEmpty() )
+  {
+    return QStringLiteral();
+  }
+
+  return QGeoCoordinate( point.x(), point.y() ).toString( QGeoCoordinate::DegreesMinutesWithHemisphere );
```
[the remark this answers is not in the document]
Comment (tomasMizera) [reply app/inpututils.cpp:246]: https://doc.qt.io/qt-5/qgeocoordinate.html example: `27° 28.054 S', 153° 1.673' E`

File: app/positionkit.h (hunk 1 above)
Comment (tomasMizera) [reply app/positionkit.h:246]: It is used in tests

**Review by PeterPetrik — APPROVED** [review]
code looks good!

**Pull request merged**